Add list of defered aggregations to the profiler

nik9000 opened the pull request:
This adds a few things to the `breakdown` of the profiler:
* `histogram` aggregations now contain `total_buckets` which is the
  count of buckets that they collected. This could be useful when
  debugging a histogram inside of another bucketing agg that is fairly
  selective.
* All bucketing aggs that can delay their sub-aggregations will now add
  a list of delayed sub-aggregations. This is useful because we
  sometimes have fairly involved logic around which sub-aggregations get
  delayed and this will save you from having to guess.
* Aggregtations wrapped in the `MultiBucketAggregatorWrapper` can't
  accurately add anything to the breakdown. Instead they the wrapper
  adds a marker entry `"multi_bucket_aggregator_wrapper": true` so we
  can be quickly pick out such aggregations when debugging.

It also fixes a bug where `_count` breakdown entries were contributing
to the overall `time_in_nanos`. They didn't add a large amount of time
so it is unlikely that this caused a big problem, but I was there.

To support the arbitrary breakdown data this reworks the profiler so
that the `breakdown` can contain any data that is supported by
`StreamOutput#writeGenericValue(Object)` and
`XContentBuilder#value(Object)`.

nik9000 committed:
Add list of defered aggregations to the profiler

This adds a few things to the `breakdown` of the profiler:
* `histogram` aggregations now contain `total_buckets` which is the
  count of buckets that they collected. This could be useful when
  debugging a histogram inside of another bucketing agg that is fairly
  selective.
* All bucketing aggs that can delay their sub-aggregations will now add
  a list of delayed sub-aggregations. This is useful because we
  sometimes have fairly involved logic around which sub-aggregations get
  delayed and this will save you from having to guess.
* Aggregtations wrapped in the `MultiBucketAggregatorWrapper` can't
  accurately add anything to the breakdown. Instead they the wrapper
  adds a marker entry `"multi_bucket_aggregator_wrapper": true` so we
  can be quickly pick out such aggregations when debugging.

It also fixes a bug where `_count` breakdown entries were contributing
to the overall `time_in_nanos`. They didn't add a large amount of time
so it is unlikely that this caused a big problem, but I was there.

To support the arbitrary breakdown data this reworks the profiler so
that the `breakdown` can contain any data that is supported by
`StreamOutput#writeGenericValue(Object)` and
`XContentBuilder#value(Object)`.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Aggregations)

nik9000 committed:
Fix name

talevy reviewed (commented):
overall looks good.

I worry about the bwc and Kibana UI. If we can be sure to mention where we stand with those things before merging, then that would be great!

nik9000 committed:
Merge branch 'master' into agg_debug_data

nik9000 committed:
Merge branch 'master' into agg_debug_data

nik9000 committed:
Update docs

nik9000 commented on server/src/main/java/org/elasticsearch/search/profile/ProfileResult.java:
```diff
@@ -49,26 +51,25 @@
 
     private final String type;
     private final String description;
-    private final Map<String, Long> timings;
```
This is the change the breaks BWC. I'm not 100% sure it is ok. I'm hunting.

The effects will be that versions of the high level rest client that don't have this won't be able to read breakdowns that include non-numbers. They'll blow up at:
```
ensureExpectedToken(XContentParser.Token.VALUE_NUMBER, parser.nextToken(), parser::getTokenLocation);
```

nik9000 replied:
Now that I think about this I'm going to move the "extra" stuff into an "extra" field and leave the breakdown all longs. I think that preserves backwards compatibility and is less confusing.

nik9000 committed:
Don't break bwc

nik9000 committed:
Finish

nik9000 committed:
Fix test

nik9000 committed:
Merge branch 'master' into agg_debug_data

nik9000 committed:
Merge branch 'master' into agg_debug_data

nik9000 committed:
Test cleanup

nik9000 commented:
@tal, this is ready for another look, I think. I've got to fix up the bwc and the test, but I think I'm mostly there.

polyfractal commented:
> I worry about the bwc and Kibana UI. 

Haven't had a chance to look at the PR itself yet, but I agree this is a potential concern.  Would someone from @elastic/es-ui be able to take a look and verify?  I don't _think_ the response changes will affect the profiler and it'll just be ignored... but not positive.

The change would be a new `"debug"` section added to aggregator profile sections (@nik9000 correct me if I'm wrong :sweat_smile:  ):

```js
            "children" : [
              {
                "type" : "LongTermsAggregator",
                "description" : "my_level_agg",
                "time_in_nanos" : 160329,
                "breakdown" : {
                  "reduce" : 0,
                  "build_aggregation" : 55712,
                  "build_aggregation_count" : 1,
                  "initialize" : 10559,
                  "initialize_count" : 1,
                  "reduce_count" : 0,
                  "collect" : 94052,
                  "collect_count" : 4
                  "collect_count" : 4,
                },
                "debug": {  // <--- this is new
                  "wrapped_in_multi_bucket_aggregator": true
                }
              }
            ]
```

nik9000 commented:
> Haven't had a chance to look at the PR itself yet, but I agree this is a potential concern. Would someone from @elastic/es-ui be able to take a look and verify? I don't _think_ the response changes will affect the profiler and it'll just be ignored... but not positive.
> 
> The change would be a new `"debug"` section added to aggregator profile sections (@nik9000 correct me if I'm wrong sweat_smile ):
> 
> ```js
> 
> ```

@polyfractal , Originally I had added the `debug` stuff under the `breakdown` section and it was kind of scary from a bwc and UI standpoint. I did talk to them about it and changed it to this and they felt fairly comfortable with it. They haven't tested it but felt like adding the new section was fairly low risk and they'd test after I merged it.

polyfractal commented:
:+1: roger, apologies for the ping noise ES-UI! :)

nik9000 commented:
> +1 roger, apologies for the ping noise ES-UI! :)

Better than not pinging them at all!

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/DeferableBucketAggregator.java:
```diff
@@ -30,10 +30,12 @@
 public abstract class DeferableBucketAggregator extends BucketsAggregator {
 
     private DeferringBucketCollector recordingWrapper;
+    private List<String> deferredAggregationNames;
```
Is there not a way to fetch the names out of `collectors` later when we need them?

nik9000 replied:
Not without some weird casting! I'll have another look just in case.

polyfractal replied:
Ah ok, no worries then :)

polyfractal reviewed (approved):
Left a question but LGTM.  Although grain of salt, I haven't looked at the profile bits of code in ages so buyer beware :)

nik9000 committed:
Merge branch 'master' into agg_debug_data

nik9000 committed:
Merge branch 'master' into agg_debug_data

nik9000 committed:
Fix after merge

nik9000 committed:
Fix test

nik9000 commented:
Thanks @talevy and @polyfractal ! I've merged and am backporting. Then I'll bump the skip in master.
